### PR TITLE
trajectory_tracker: fix local goal handling

### PR DESCRIPTION
**Pull request by at-wat**

### CODE REVIEW

**Comment by at-wat** [comment]
## [#713] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_costmap_watchdog_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_costmap_watchdog.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_path2d.xml: 2 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker.xml: 3 tests
Summary: 64 tests, 0 errors, 0 failures, 0 skipped
```
</details>

**Comment by at-wat** [comment]
## [#717] FAILED on indigo

```catkin_make``` failed

**Comment by at-wat** [comment]
## before refactoring
trajectory_tracker_tests

## after refactoring
trajectory_tracker_tests_latest

## this PR
trajectory_tracker_tests_fixed

**Comment by at-wat** [comment]
## [#718] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_costmap_watchdog_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_costmap_watchdog.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_path2d.xml: 2 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker.xml: 5 tests
Summary: 66 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: trajectory_tracker/include/trajectory_tracker/path2d.h [hunk 1]
```diff
@@ -93,7 +93,7 @@ class Path2D : public std::vector<Pose2D>
       const float angle_pose = allow_backward_motion ? it->yaw_ : angle;
       const float sign_vel_req = std::cos(angle) * std::cos(angle_pose) + std::sin(angle) * std::sin(angle_pose);
       if (sign_vel_prev * sign_vel_req < 0)
-        return it + 1;
```
Comment (at-wat) [inline trajectory_tracker/include/trajectory_tracker/path2d.h]: `it` is already next to the switching back point.

File: trajectory_tracker/src/trajectory_tracker.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ void TrackerNode::control()
 
   // Find nearest line strip
   const trajectory_tracker::Path2D::ConstIterator it_local_goal =
-      lpath.findLocalGoal(lpath.begin(), lpath.end(), allow_backward_);
+      lpath.findLocalGoal(lpath.begin() + path_step_done_, lpath.end(), allow_backward_);
```
Comment (at-wat) [inline trajectory_tracker/src/trajectory_tracker.cpp]: must search next local goal point

File: trajectory_tracker/src/trajectory_tracker.cpp [hunk 3]
```diff
@@ -361,8 +369,10 @@ void TrackerNode::control()
   {
     if (path_length < min_track_path_ || std::abs(remain_local) < stop_tolerance_dist_)
     {
-      angle = trajectory_tracker::angleNormalized(-lpath.back().yaw_);
+      angle = trajectory_tracker::angleNormalized(-(it_local_goal - 1)->yaw_);
```
Comment (at-wat) [inline trajectory_tracker/src/trajectory_tracker.cpp]: this bug had been existed before the refactoring

File: trajectory_tracker/src/trajectory_tracker.cpp [hunk 4]
```diff
@@ -443,9 +458,11 @@ void TrackerNode::control()
   tracking.pose.orientation = tf2::toMsg(tf2::Quaternion(tf2::Vector3(0.0, 0.0, 1.0), -angle));
   pub_tracking_.publish(tracking);
 
-  path_step_done_ = i_nearest - 2;  // i_nearest is the next of the nearest node
+  path_step_done_ = i_nearest - 1;  // i_nearest is the next of the nearest node
```
Comment (at-wat) [inline trajectory_tracker/src/trajectory_tracker.cpp]: next goal and nearest point are searched from `path_step_done_`. `- 1` is enough to search from current nearest point.

**Comment by at-wat** [comment]
@ykoga-kyutech @DaikiMaekawa PTAL

File: trajectory_tracker/include/trajectory_tracker/path2d.h (hunk 1 above)
Comment (at-wat) [reply trajectory_tracker/include/trajectory_tracker/path2d.h]: It's also covered by TrajectoryTrackerTest/SwitchBack.
Adding separate unit test for it is good.

**Comment by at-wat** [comment]
## [#721] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_costmap_watchdog_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_costmap_watchdog.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_path2d.xml: 4 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_tracker_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_tracker.xml: 5 tests
Summary: 68 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: trajectory_tracker/src/trajectory_tracker.cpp (hunk 3 above)
Comment (DaikiMaekawa) [reply trajectory_tracker/src/trajectory_tracker.cpp]: got it. thanks.

**Pull request merged**